File: src/zork/ZorkActions.java
package zork;

import java.util.Queue;
import zork.enums.EItem;
import zork.enums.ENpc;
import zork.enums.EStats;
import zork.enums.EVerbs;
import zork.models.Room;
import zork.models.entities.Npc;
import zork.utils.Parser;

/**
 * Here the game actions are defined. Only this class should know the state of
 * the game objects
 *
 * @author d.peters
 * @version 1.0
 */
public class ZorkActions {
  private final Parser parser;
  private final GameState game;

  public ZorkActions(Parser parser, GameState game) {
    this.parser = parser;
    this.game = game;
  }

  /**
   * Ask the user to specify if "go" was entered without any direction and/or
   * go towards the specified direction.
   *
   * @param commands the user command
   */
  public void specifyDirection(Queue<EVerbs> commands) {
    EVerbs command;

    if (commands.isEmpty()) {

      System.out.println("\nWhere do you want to go?");
      commands = this.parser.getCommand();

    }
    command = commands.poll();
    Room nextRoom = this.game.getCurrentRoom().goToNextRoom(command);

    // Check for invalid directions
    if (nextRoom != null) {
      this.game.setCurrentRoom(nextRoom);
    } else {
      System.out.println("\nYou walked into a wall...");
    }
  }

  /**
   * Add item from room to inventory, if it exists. If no item was specified,
   * then ask user to specify.
   * TODO: Allow picking up single items instead of just all
   *
   * @param commands user commands
   */
  public void specifyObject(Queue<EVerbs> commands) {
    EVerbs command;

    if (commands.isEmpty()) {

      System.out.println("\nTake what?");
      commands = this.parser.getCommand();

    }
    command = commands.poll();

    if (command.equals(EVerbs.ALL)) {

      this.game.getCurrentRoom().getItems().forEach((item) -> {
        System.out.println(
            "\n" + item.getName() + " added to your inventory."
        );

        this.game.getPlayer().getInventory().add(item);
      });
      this.game.getCurrentRoom().getItems().clear();
    }
  }

  /**
   * Use item specified or ask user to name item to be used.
   *
   * @param commands parsed user commands
   */
  public void specifyItem(Queue<EVerbs> commands) {
    EVerbs command;

    if (commands.isEmpty()) {

      System.out.println("\nUse what?");
      commands = this.parser.getCommand();

    }
    command = commands.poll();

    if (command.equals(EVerbs.ALL)) {
      this.game.getPlayer().getInventory().forEach(item -> item.use(this.game.getPlayer()));
    } else {
      for (EItem item : EItem.values()) {
        if (EItem.valueOf(command.toString()).equals(item)) {
          this.game.getPlayer().getInventory().forEach(itm -> {
            if (itm.getName().toUpperCase().equals(item.toString())) {
              itm.use(this.game.getPlayer());
            }
          });
        }
      }
    }
  }

  /**
   * Talk to an npc. Ask for dialog partner if none given.
   *
   * @param commands user submitted commands
   */
  public void talkToNpc(Queue<EVerbs> commands) {
    EVerbs command;

    if (commands.isEmpty()) {
      System.out.println("\nTo whom?");
      commands = this.parser.getCommand();
    }

    command = commands.poll();

    if (!this.game.getCurrentRoom().getCharacters().isEmpty()) {
      for (ENpc item : ENpc.values()) {
        if (ENpc.valueOf(command.toString()).equals(item)) {
          ENpc npcId = ENpc.valueOf(command.toString());
          this.game.getCurrentRoom().getCharacters().forEach(character -> {
            if (character instanceof Npc) {
              Npc npc = (Npc) character;
              if (npc.getId().equals(npcId)) {
                this.game.getPlayer().talkTo(npc);
              }
            }
          });
        }
      }
    }

  }

  /**
   * Let the player hit himself if requested.
   */
  public void masochistAction() {
    double selfDamage = this.game.getPlayer().getStats().get(EStats.ATTACK).getValue();
    System.out.println("\nYou hit yourself with a bludgeon.");
    this.game.getPlayer().takeDamage(selfDamage);
  }

  /**
   * Attack a character.
   *
   * @param commands user submitted commands
   */
  public void attackAction(Queue<EVerbs> commands) {

    EVerbs command;

    if (commands.isEmpty()) {

      System.out.println("\nAttack what?");
      commands = this.parser.getCommand();

    }

    command = commands.poll();

    if (!this.game.getCurrentRoom().getCharacters().isEmpty()) {

      for (ENpc item : ENpc.values()) {
        if (ENpc.valueOf(command.toString()).equals(item)) {
          ENpc npcId = ENpc.valueOf(command.toString());
          game.getCurrentRoom().getCharacters().forEach(character -> {
            if (character instanceof Npc) {
              Npc npc = (Npc) character;
              if (npc.getId().equals(npcId)) {
                npc.takeDamage(game.getPlayer().getStats().get(EStats.ATTACK).getValue());
              }
            }
          });
        }
      }
    }
  }

  /**
   * Tell the player if there are any NPC around.
   */
  public void foundNpc() {
    if (!game.getCurrentRoom().getCharacters().isEmpty()) {
      game.getCurrentRoom().getCharacters().forEach(character -> {
        if (character.isAlive()) {
          System.out.println(character.getName() + " is in the room...");
        } else {
          System.out.println("You sense the corpse of " + character.getName());
        }
      });

    }
  }

  /**
   * Message printed when the player starts the game.
   */
  public void welcomeMessage() {

    System.out.println("Hail brave " + game.getPlayer().getName() + ". Welcome to Zork.");

  }

  /**
   * Message printed when the player quits the game.
   */
  public void goodbyeMessage() {
    System.out.println("\nSee you again soon " + game.getPlayer().getName() + "!");
  }

  /**
   * List the commands available.
   */
  public void showHelp() {

    for (EVerbs c : EVerbs.values()) {
      System.out.println(c + " - " + c.getDetails());
    }

  }

  /**
   * Print the description of the room.
   */
  public void roomDescription() {
    System.out.println(game.getCurrentRoom().getDescription());
  }

  /**
   * Tell the player his current location.
   */
  public void currentRoomMessage() {
    System.out.println("\nYou are in the " + game.getCurrentRoom().getName() + ".");
  }

  /**
   * Send message if user gave an invalid command.
   */
  public void unknownCmdMsg() {
    System.out.println("\nI don't understand that command.");
  }

  /**
   * Ask the player to do something.
   */
  public void askForAction() {
    System.out.println("What do you want to do?");
  }

  /**
   * Get commands from parser.
   *
   * @return new command from parser
   */
  public Queue<EVerbs> getCommand() {
    return parser.getCommand();
  }

  /**
   * Print contents of the players inventory.
   */
  public void showPlayerInventory() {
    game.getPlayer().showInventory();
  }

  /**
   * Print player stats.
   */
  public void showPlayerStats() {
    game.getPlayer().displayStats();
  }

  /**
   * Check if player is alive.
   *
   * @return boolean flag for alive
   */
  public boolean playerIsAlive() {
    return this.game.getPlayer().isAlive();
  }
}

File: src/zork/application/ZorkActions.kt
package zork.application

import java.util.Queue
import zork.enums.EItem
import zork.enums.ENpc
import zork.enums.EStats
import zork.enums.EVerbs
import zork.interfaces.IUserInterface
import zork.models.Room
import zork.models.entities.Npc
import zork.utils.Parser

/**
 * Here the game actions are defined. Only this class should know the state of
 * the game objects
 *
 * @author d.peters
 * @version 1.0
 */
class ZorkActions(private val parser: Parser, private val ui: IUserInterface, private val game: GameState) {
  /**
   * Ask the user to specify if "go" was entered without any direction and/or
   * go towards the specified direction.
   *
   * @param commands the user command
   */
  fun specifyDirection(commands: Queue<EVerbs>) {
    var commandsList = commands
    if (commandsList.isEmpty()) {
      ui.displayMessage("Where do you want to go?")
      commandsList = parser.getCommand()
    }
    val command = commandsList.poll()
    val nextRoom = game.currentRoom.goToNextRoom(command)

    // Check for invalid directions
    if (nextRoom != Room.wall) {
      game.currentRoom = nextRoom
    } else {
      ui.displayMessage("You walked into a wall...")
    }
  }

  /**
   * Add item from room to inventory, if it exists. If no item was specified,
   * then ask user to specify.
   * TODO: Allow picking up single items instead of just all
   *
   * @param commands user commands
   */
  fun specifyObject(commands: Queue<EVerbs>) {
    var commandsList = commands

    if (commandsList.isEmpty()) {
      ui.displayMessage("Take what?")
      commandsList = parser.getCommand()
    }
    val command = commandsList.poll()

    if (command == EVerbs.ALL) {
      game.currentRoom.items.forEach({ item ->
        ui.displayMessage(item.name + " added to your inventory.")
        game.player.inventory.add(item)
      })
      game.currentRoom.items.clear()
    }
  }

  /**
   * Use item specified or ask user to name item to be used.
   *
   * @param commands parsed user commands
   */
  fun specifyItem(commands: Queue<EVerbs>) {
    var commandsList = commands
    if (commands.isEmpty()) {
      ui.displayMessage("Use what?")
      commandsList = parser.getCommand()
    }
    val command = commandsList.poll()

    if (command == EVerbs.ALL) {
      game.player.inventory.forEach { item -> item.use() }
    } else {
      EItem.values().forEach { item ->
        if (EItem.valueOf(command.toString()) == item) {
          game.player.inventory.forEach { itm ->
            if (itm.name.toUpperCase() == item.toString()) {
              itm.use()
            }
          }
        }
      }
    }
  }

  /**
   * Talk to an npc. Ask for dialog partner if none given.
   *
   * @param commands user submitted commands
   */
  fun talkToNpc(commands: Queue<EVerbs>) {
    var commandsList = commands
    if (commandsList.isEmpty()) {
      ui.displayMessage("To whom?")
      commandsList = parser.getCommand()
    }

    val command = commandsList.poll()

    if (!game.currentRoom.characters.isEmpty()) {
      ENpc.values().forEach {
        val npcId = ENpc.valueOf(command.toString())
        game.currentRoom.characters.forEach { character ->
          if (character is Npc) {
            if (character.npcId == npcId) {
              game.player.talkTo(character)
            }
          }
        }
      }
    }
  }


  /**
   * Let the player hit himself if requested.
   */
  fun masochistAction() {
    val selfDamage = game.player.stats[EStats.ATTACK]!!.value
    ui.displayMessage("You hit yourself with a bludgeon.")
    game.player.takeDamage(selfDamage)
  }

  /**
   * Attack a character.
   *
   * @param commands user submitted commands
   */
  fun attackAction(commands: Queue<EVerbs>) {
    var commandsList = commands

    if (commandsList.isEmpty()) {
      ui.displayMessage("Attack what?")
      commandsList = parser.getCommand()

    }

    val command = commandsList.poll()

    if (!game.currentRoom.characters.isEmpty()) {
      ENpc.values().forEach { item ->
        if (ENpc.valueOf(command.toString()) == item) {
          val npcId = ENpc.valueOf(command.toString())
          game.currentRoom.characters.forEach({ character ->
            if (character is Npc) {
              if (character.npcId == npcId) {
                character.takeDamage(game.player.stats[EStats.ATTACK]!!.value)
              }
            }
          })
        }
      }
    }
  }

  /**
   * Tell the player if there are any NPC around.
   */
  fun foundNpc() {
    if (!game.currentRoom.characters.isEmpty()) {
      game.currentRoom.characters.forEach({ character ->
        if (character.alive) {
          ui.displayMessage(character.name + " is in the room...")
        } else {
          ui.displayMessage("You sense the corpse of " + character.name)
        }
      })

    }
  }

  /**
   * Message printed when the player starts the game.
   */
  fun welcomeMessage() {
    ui.displayMessage("Hail brave " + game.player.name + ". Welcome to Zork.")
  }

  /**
   * Message printed when the player quits the game.
   */
  fun goodbyeMessage() {
    ui.displayMessage("See you again soon " + game.player.name + "!")
  }

  /**
   * List the commands available.
   */
  fun showHelp() {
    EVerbs.values().forEach { c -> ui.displayMessage(c.toString() + " - " + c.details) }
  }

  /**
   * Print the description of the room.
   */
  fun roomDescription() {
    ui.displayMessage(game.currentRoom.description)
  }

  /**
   * Tell the player his current location.
   */
  fun currentRoomMessage() {
    ui.displayMessage("You are in the " + game.currentRoom.name + ".")
  }

  /**
   * Send message if user gave an invalid command.
   */
  fun unknownCmdMsg() {
    ui.displayMessage("I don't understand that command.")
  }

  /**
   * Ask the player to do something.
   */
  fun askForAction() {
    ui.displayMessage("What do you want to do?")
  }

  /**
   * Get commands from parser.
   *
   * @return new command from parser
   */
  fun getCommand(): Queue<EVerbs> {
    return parser.getCommand()
  }

  /**
   * Print contents of the players inventory.
   */
  fun showPlayerInventory() {
    game.player.showInventory()
  }

  /**
   * Print player stats.
   */
  fun showPlayerStats() {
    game.player.displayStats()
  }

  /**
   * Check if player is alive.
   *
   * @return boolean flag for alive
   */
  fun playerIsAlive(): Boolean {
    return game.player.alive
  }
}
